### PR TITLE
[h264d]fix reference frame mark to support discontinuous frame_num

### PR DESCRIPTION
**Pull request by chuanli1**

### CODE REVIEW

**Comment by wangyan-intel** [comment]
@chuanli1 Could you please add the detail info into commit message like you mentioned in #2452?  You could make summary in commit message. And add one line like "Fixes #2452." which will close the issue automatically when your PR is merged. Thanks.

**Review by dmitryermilov — CHANGES_REQUESTED** [review]
Does the change follow the spec? Please provide a link or wording from AVC spec which say that in case of available ref frames we need to skip frame gaps handling.

**Comment by dmitryermilov** [comment]
@dgurulev , could you review as well?

**Comment by chuanli1** [comment]
thanks @dmitryermilov and @wangyan-intel . 
Rec. ITU-T H.264 pointed out in the Decoding process for gaps in frame_num section that the bitstream shall not contain data that result in the derivation of a reference picture that is marked as "non-existing" in any invocation of the reference picture selection process specified in clause 8.4.2.1.

In order to cover this kind of cases，I judge whether there have the mark reference frames according to count NumShortTermRefs and NumLongTermRefs, if yes, current frame will not use a non-existing frame as reference frame and continue to use the already  marked reference frame.

**Comment by chuanli1** [comment]
@dmitryermilov  this CI build is the environment issue, and there is no relationship with my patch.

**Comment by dmitryermilov** [comment]
> @dmitryermilov this CI build is the environment issue, and there is no relationship with my patch.

@eponomar , can you please help?

**Comment by eponomar** [comment]
> > @dmitryermilov this CI build is the environment issue, and there is no relationship with my patch.
> 
> @eponomar , can you please help?

Checks restarted. It seems it was local issue or force push to driver master.

File: _studio/shared/umc/codec/h264_dec/src/umc_h264_task_supplier.cpp [hunk 1]
```diff
@@ -2729,7 +2729,12 @@ Status TaskSupplier::ProcessFrameNumGap(H264Slice *pSlice, int32_t field, int32_
     Status umcRes = UMC_OK;
     const H264SeqParamSet* sps = pSlice->GetSeqParam();
     H264SliceHeader *sliceHeader = pSlice->GetSliceHeader();
-
+	uint32_t NumShortTermRefs, NumLongTermRefs;
+	view.GetDPBList(0)->countActiveRefs(NumShortTermRefs, NumLongTermRefs);
+	if (NumShortTermRefs > 0 || NumLongTermRefs > 0) 
+	{
+		return UMC_OK;
```
Comment (dgurulev) [inline _studio/shared/umc/codec/h264_dec/src/umc_h264_task_supplier.cpp:2736]: Here u even don't try to check (see `DetectFrameNumGap`) whether we have gaps or not (so actually throw away all function's code below). The second thing, the behavior of the decoder when [SPS :: gaps_in_frame_num_value_allowed_flag] is well defined (see 8.2.5.2).

**Review by dgurulev — CHANGES_REQUESTED** [review]

**Comment by chuanli1** [comment]
As 8.2.5.2 mentioned, filling the frame_num gap with non-existing frames and canceling the previously marked reference frame, and finally marking the non-existent frame as a reference frame, the result will appear a green screen , I think this is a right phenomenon, because the purpose of filling the non-existent frame is that the current frame cannot find the reference frame, but in order to cover the clip, before inserting this non-existent frame, first check whether there is a marked reference frame.

File: _studio/shared/umc/codec/h264_dec/src/umc_h264_task_supplier.cpp (hunk 1 above)
Comment (dmitryermilov) [reply _studio/shared/umc/codec/h264_dec/src/umc_h264_task_supplier.cpp:2736]: @chuanli1 , yes, it seems the change is not according to the spec. By “if (NumShortTermRefs > 0 || NumLongTermRefs > 0)” you disable all further actions even if gaps_in_frame_num_value_allowed_flag is allowed. Which is not right. However even if input stream is corrupted we need to at least try to reach better error tolerances in the decoder.  I mean, if possible we need to try to find a WA which on the one hand doesn’t contradict the spec, and on the other hand resolves the issue.

BTW, could you check behavior with other decoders (reference JM, libav, vaapi, something else)?

**Review by dmitryermilov — CHANGES_REQUESTED** [review]

**Comment by wangyan-intel** [comment]
@chuanli1 Have you done something wrong？ Your commit is pushed another PR?

**Comment by chuanli1** [comment]
sorry, submit wrong.

**Pull request closed without merging**